Release GIL for RPC pybind functions.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33610 Release GIL for RPC pybind functions.**

Our pybind definitions for several RPC functions didn't release GIL
once we were processing stuff in C++.

This PR adds asserts that we release GIL appropriately and adds
py::gil_scoped_release and py::gil_scoped_acquire in the appropriate places.

Differential Revision: [D20025847](https://our.internmc.facebook.com/intern/diff/D20025847/)

pritamdamania committed:
Release GIL for RPC pybind functions.

Our pybind definitions for several RPC functions didn't release GIL
once we were processing stuff in C++.

This PR adds asserts that we release GIL appropriately and adds
py::gil_scoped_release and py::gil_scoped_acquire in the appropriate places.

Differential Revision: [D20025847](https://our.internmc.facebook.com/intern/diff/D20025847/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Release GIL for RPC pybind functions."


Our pybind definitions for several RPC functions didn't release GIL
once we were processing stuff in C++.

This PR adds asserts that we release GIL appropriately and adds
py::gil_scoped_release and py::gil_scoped_acquire in the appropriate places.

Differential Revision: [D20025847](https://our.internmc.facebook.com/intern/diff/D20025847/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Release GIL for RPC pybind functions."


Our pybind definitions for several RPC functions didn't release GIL
once we were processing stuff in C++.

This PR adds asserts that we release GIL appropriately and adds
py::gil_scoped_release and py::gil_scoped_acquire in the appropriate places.

Differential Revision: [D20025847](https://our.internmc.facebook.com/intern/diff/D20025847/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Release GIL for RPC pybind functions."


Our pybind definitions for several RPC functions didn't release GIL
once we were processing stuff in C++.

This PR adds asserts that we release GIL appropriately and adds
py::gil_scoped_release and py::gil_scoped_acquire in the appropriate places.

Differential Revision: [D20025847](https://our.internmc.facebook.com/intern/diff/D20025847/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Release GIL for RPC pybind functions."


Our pybind definitions for several RPC functions didn't release GIL
once we were processing stuff in C++.

This PR adds asserts that we release GIL appropriately and adds
py::gil_scoped_release and py::gil_scoped_acquire in the appropriate places.

Differential Revision: [D20025847](https://our.internmc.facebook.com/intern/diff/D20025847/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -27,6 +27,8 @@ std::shared_ptr<Operator> matchBuiltinOp(
     const py::kwargs& kwargs,
     Stack& stack) {
+  // Acquire GIL for py::args and py::kwargs processing.
+  pybind11::gil_scoped_acquire ag;
```
do we need the GIL here or can we move it to wrap `createStackForSchema `?

pritamdamania87 replied:
I put it up here since py::args and py::kwargs are used in the VLOG and in the TORCH_CHECK at the end. Those statements need to be inside the GIL as well.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -302,17 +303,21 @@ If the future completes with an error, an exception is thrown.
          const std::shared_ptr<torch::autograd::profiler::RecordFunction>& rf,
          const py::args& args,
          const py::kwargs& kwargs) {
+        DCHECK(!PyGILState_Check());
```
Just want to make sure that I understand this correctly. The behavior is if the GIL is held at this point, it will crash immediately?

satgera replied:
I believe the behaviour is that if the GIL is held by "this thread" then it will crash.

pritamdamania87 replied:
This would crash only in "debug" mode. I used DCHECK to avoid any perf overheads in prod. In prod, this check wouldn't even execute.

zhaojuanmao commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -378,22 +389,30 @@ If the future completes with an error, an exception is thrown.
+          stack = torch::jit::createStackForSchema(
+              functionSchema, args, kwargs, c10::nullopt);
+        }
         auto rrefPtr = remoteTorchscript(
```
nit: add check before remoteTorchscript()

zhaojuanmao commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -363,8 +369,13 @@ If the future completes with an error, an exception is thrown.
+          stack = torch::jit::createStackForSchema(
+              functionSchema, args, kwargs, c10::nullopt);
+        }
         auto fut =
```
nit: add check before rpcTorchscript()

mrshenli reviewed (approved):
LGTM! Thanks for fixing!

zhaojuanmao reviewed (approved):
lgtm, some nits

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -27,6 +27,8 @@ std::shared_ptr<Operator> matchBuiltinOp(
+  // Acquire GIL for py::args and py::kwargs processing.
+  pybind11::gil_scoped_acquire ag;
   Symbol symbol = Symbol::fromQualString(opName);
   if (symbol.is_aten()) {
```
Since `matchBuiltinOp(..)` is called in both `pyRpcBuiltin` and `pyRemoteBuiltin`, will this incur Python thread context switch cost for these two entries? 
It's releases GIL on pybind11 call, but after pybind11 call dispatch, it immediately acquires GIL.

Is it better if we release GIL after the call to `matchBuiltinOp(..)`?

```cpp
  module.def(
      "_invoke_rpc_builtin",
      [](const WorkerInfo& dst,
         const std::string& opName,
         const std::shared_ptr<torch::autograd::profiler::RecordFunction>& rf,
         const py::args& args,
         const py::kwargs& kwargs) {
        return pyRpcBuiltin(dst, opName, rf, args, kwargs);
      }, py::call_guard<gil_scoped_acquire>);
```

```cpp
std::shared_ptr<FutureMessage> pyRpcBuiltin(
    const WorkerInfo& dst,
    const std::string& opName,
    const std::shared_ptr<torch::autograd::profiler::RecordFunction>& rf,
    const py::args& args,
    const py::kwargs& kwargs) {
  Stack stack;
  auto op = matchBuiltinOp(opName, args, kwargs, stack);
  py::gil_scoped_release release;
  auto scriptCall = std::make_unique<ScriptCall>(op, std::move(stack));
}
```

pritamdamania87 replied:
Good point, I'll reorder this.

pritamdamania committed:
Update on "Release GIL for RPC pybind functions."


Our pybind definitions for several RPC functions didn't release GIL
once we were processing stuff in C++.

This PR adds asserts that we release GIL appropriately and adds
py::gil_scoped_release and py::gil_scoped_acquire in the appropriate places.

Differential Revision: [D20025847](https://our.internmc.facebook.com/intern/diff/D20025847/)

[ghstack-poisoned]

xush6528 reviewed (approved):
Looks perfect now. Thanks!
